src/mozilla_sec_eia/ex_21/inference.py

```diff
@@ -0,0 +1,366 @@
+
+    def preprocess(self, example):
+        """Encode and tokenize model inputs."""
+        image = example["image"]
+        words = example["tokens"]
+        boxes = example["bboxes"]
+        encoding = self.tokenizer(
+            image,
+            words,
+            boxes=boxes,
+            return_tensors="pt",
+            truncation=True,
+            padding="max_length",
+            max_length=512,  # this is the maximum max_length
+            stride=128,
+            return_offsets_mapping=True,
+            return_overflowing_tokens=True,
+        )
+        model_inputs = {}
+        model_inputs["raw_encoding"] = encoding.copy()
+        model_inputs["example"] = example
+        model_inputs["offset_mapping"] = encoding.pop("offset_mapping")
+        model_inputs["sample_mapping"] = encoding.pop("overflow_to_sample_mapping")
+        # TODO: do we actually need to make these into ints?
+        encoding["input_ids"] = encoding["input_ids"].to(torch.int64)
+        encoding["attention_mask"] = encoding["attention_mask"].to(torch.int64)
+        encoding["bbox"] = encoding["bbox"].to(torch.int64)
+        encoding["pixel_values"] = torch.stack(encoding["pixel_values"])
+        model_inputs["encoding"] = encoding
+        return model_inputs
+
+    def _forward(self, model_inputs):
+        encoding = model_inputs["encoding"]
+        if torch.cuda.is_available():
+            encoding.to("cuda")
+            self.model.to("cuda")
+        # since we're doing inference, we don't need gradient computation
+        with torch.no_grad():
+            output = self.model(**encoding)
+            return {
+                "logits": output.logits,
+                "predictions": output.logits.argmax(-1).squeeze().tolist(),
+                "raw_encoding": model_inputs["raw_encoding"],
+                "example": model_inputs["example"],
+            }
+
+    def postprocess(self, all_outputs):
+        """Return logits, model predictions, and the extracted dataframe."""
+        logits = all_outputs["logits"]
+        predictions = all_outputs["logits"].argmax(-1).squeeze().tolist()
+        output_df = self.extract_table(all_outputs)
+        return logits, predictions, output_df
+
+    def extract_table(self, all_outputs):
+        """Extract a structured table from a set of inference predictions.
+
+        This function essentially works by stacking bounding boxes and predictions
+        into a dataframe and going from left to right and top to bottom. Then, every
+        every time a new subsidiary entity is encountered, it assigns a new group or
+        "row" to that subsidiary. Next, location and ownership percentage words/labeled
+        entities in between these subsidiary groups are assigned to a subsidiary row/group.
+        Finally, this is all formatted into a dataframe with an ID column from the original
+        filename and a basic cleaning function normalizes strings.
+        """
```

Comment:
At some point it might be nice to split this into some sub-functions to make it clearer what's doing what. Probably something we can hold off on until we feel like the model is a bit more mature though

Reply:
Made a TODO comment to do this once the model is more stable.